### PR TITLE
Memory Tracker for tracking Module wise memory

### PR DESCRIPTION
**Pull request by sanketpurandare**
A tool for measuring module wise memory consumption. It also provides a breakdown of memory consumption in terms of parameters, gradients, activations and optimizer state. Works in fake tensor mode as well. 
 
cc: @albanD @Chillee @wanchaol @wconstab @gnadathur 

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @LucasLLC

### CODE REVIEW

File: torch/utils/mem_tracker.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.utils._pytree import tree_map_only
+import torchvision
+import torch.nn as nn
+from torch.utils._python_dispatch import TorchDispatchMode
+from torch.utils.weak import WeakIdKeyDictionary
+import weakref
+from enum import Enum
+
+MB = 2**20
```
Comment (albanD) [inline torch/utils/mem_tracker.py:15]: Since this is a public namespace, we need to make private everything that we don't want to expose to the user.
The two options here are:
- Make everything that is private start with a `_`
- Add a `__all__` attribute that lists the public APIs

File: torch/utils/mem_tracker.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+KB = 2**10
+MEMORY_MAX = 0
+PYTORCH_MIN_ALLOCATE = 2**9
+MEMORY_USE = WeakIdKeyDictionary()
```
Comment (albanD) [inline torch/utils/mem_tracker.py:19]: I think now that we're making this a real thing, we should remove this global and just create it on entry to the mode.

File: torch/utils/mem_tracker.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    return f
+
+
+def final_call():
+    fqn = get_fqn()
+    memory_tracking[fqn]["After Backward"] = get_current_memory_allocated()
+    parents.pop()
+
+
+def register_hooks(name: str, module: nn.Module):
```
Comment (albanD) [inline torch/utils/mem_tracker.py:155]: We have something similar for flop_counter here
Can we re-use it https://github.com/pytorch/pytorch/blob/d1da9cc654b4560bebb96f2b82923b7cab2e761e/torch/utils/flop_counter.py#L380 ?

Also we have class level hooks at https://pytorch.org/docs/main/generated/torch.nn.modules.module.register_module_forward_hook.html that will trigger on every single Module. This can be useful if you want to trigger the hook on every single module that runs without the user having to give the Module to you to instrument. Or as a way to automatically trigger the instrumentation the first time you see such a Module.

File: torch/utils/mem_tracker.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+def display_mem_stats():
+    for mod in memory_tracking.keys():
+        print(f"Module: ", mod)
+        for k, stats in memory_tracking[mod].items():
+            print(f"{k}")
+            print_mem_stats(stats)
+        print()
+
+
+def experiment():
```
Comment (albanD) [inline torch/utils/mem_tracker.py:219]: We should move this to a test file like the flop_counter.
Also we should check there that:
- The measured memory matched max_allocated to an epsilon
- It does run smoothly in fake mode and gives the same result as in real mode

**Review by albanD — COMMENTED** [review]
Thanks for sending a PR for this.
cc @Chillee any feedback on this? Also I was looking for the doc page for the flop_counter and can't find it. Where is it?

cc @fmassa this might be useful to measure the peak memory usage of a Module on cuda without having to run it for real when combined with FakeTensor mode

File: torch/utils/mem_tracker.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+KB = 2**10
+MEMORY_MAX = 0
+PYTORCH_MIN_ALLOCATE = 2**9
+MEMORY_USE = WeakIdKeyDictionary()
+FIRST_OPT_ITER = True
+parents = []
+memory_tracking = defaultdict(lambda: defaultdict(lambda: defaultdict()))
+
+
+class RefType(str, Enum):
```
Comment (Skylion007) [inline torch/utils/mem_tracker.py:25]:
```suggestion
class RefType(StrEnum):
```

File: torch/utils/mem_tracker.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+from enum import Enum
+
+MB = 2**20
+KB = 2**10
+MEMORY_MAX = 0
+PYTORCH_MIN_ALLOCATE = 2**9
+MEMORY_USE = WeakIdKeyDictionary()
+FIRST_OPT_ITER = True
+parents = []
+memory_tracking = defaultdict(lambda: defaultdict(lambda: defaultdict()))
```
Comment (Skylion007) [inline torch/utils/mem_tracker.py:22]:
```suggestion
memory_tracking = defaultdict(lambda: defaultdict(defaultdict))
```

File: torch/utils/mem_tracker.py (hunk 5 above)
Comment (sanketpurandare) [reply torch/utils/mem_tracker.py:25]: StrEnum requires an external package strenum. It's not included in pytorch requirements

File: torch/utils/mem_tracker.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, Dict
+import torch
+import math
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.utils._pytree import tree_map_only
+import torchvision
```
[the remark this answers is not in the document]
Comment (sanketpurandare) [reply torch/utils/mem_tracker.py]: Done

File: torch/utils/mem_tracker.py (hunk 1 above)
Comment (sanketpurandare) [reply torch/utils/mem_tracker.py]: Done

File: torch/utils/mem_tracker.py (hunk 2 above)
Comment (sanketpurandare) [reply torch/utils/mem_tracker.py]: Done

File: torch/utils/mem_tracker.py (hunk 3 above)
Comment (sanketpurandare) [reply torch/utils/mem_tracker.py]: Used something similar to flop counter

File: torch/utils/mem_tracker.py (hunk 6 above)
Comment (sanketpurandare) [reply torch/utils/mem_tracker.py]: Done

File: torch/utils/mem_tracker.py (hunk 4 above)
Comment (sanketpurandare) [reply torch/utils/mem_tracker.py:253]: @albanD where is this file?

Comment (albanD) [reply torch/utils/mem_tracker.py:253]: https://github.com/pytorch/pytorch/blob/f9b2ffa7c494cd1a0481149cb4089da4b33e4b66/test/test_flop_counter.py

**Comment by lezcano** [comment]
Bad rebase tagged too many people. Please open a new PR.
